text/0032-fog-compatible-gift-codes.md

```diff
@@ -0,0 +1,201 @@
+
+```
+pub struct InputCredentials {
+    /// A "ring" containing "mixins" and the one "real" TxOut to be spent.
+    pub ring: Vec<TxOut>,
+
+    /// Proof that each TxOut in `ring` is in the ledger.
+    pub membership_proofs: Vec<TxOutMembershipProof>,
+
+    /// Index in `ring` of the "real" output being spent.
+    pub real_index: usize,
+
+    /// Private key for the "real" output being spent.
+    pub onetime_private_key: RistrettoPrivate,
+
+    /// Public key of the transaction that created the "real" output being
+    /// spent.
+    pub real_output_public_key: RistrettoPublic,
+
+    /// View private key for the address this input was sent to
+    pub view_private_key: RistrettoPrivate,
+}
+```
+
+All the clients are capable of sampling rings, obtaining membership proofs, and the
+`TxOutGiftCode` contains the one-time private key.
+
+However, the `InputCredentials` appears to require the view private key, which the recipient
+of the `TxOutGiftCode` does not have.
+
+Fortunately, it turns out that they don't really need it, it is only used in the `TransactionBuilder`
+to construct the `shared_secret`:
+
+https://github.com/mobilecoinfoundation/mobilecoin/blob/924abcff4937fced12cdd951aad598dbdbbfab05/transaction/std/src/transaction_builder.rs#L410
+
+We propose either to
+* Modify `InputCredentials` so that it includes the `shared_secret` instead of the `view_private_key`.
+  Since clients usually compute the `shared_secret` as soon as they receive an input as a by-product of view-key matching,
+  this may actually be more efficient and prevent computing elliptic curve operations again unnecessarily.
+* Modify `InputCredentials` so that it holds either the `shared_secret` OR the `view_private_key`, for backwards
+  compatibility.
+
+Clients will then be able to go from a `TxOutGiftCode` to an `InputCredentials` very easily.
+
+Most likely, we could make this schema a new option to `PrintableWrapper` and deprecate the older `TransferPayload` to
+discourage continued use, since they degrade the privacy of Fog users.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+In our opinion, there is no reason why we should not do this, and this version of gift codes is superior to the existing version.
+The existing version is very complicated and forces all the clients to view-key scan the ledger with a new set of keys in order
+to process a gift code. The "fixed" version that returns the `TxOut` public key avoids the view key scanning issue, but it still
+isn't really compatible with Fog's privacy model. It is much simpler to just give away ownership of an existing `TxOut`, and
+give the fog recipient a global `TxOut` index.
+
+This also makes creating gift codes instantaneous and not require on-chain events and fees.
```

Comment:
This is true only if you have a TxOut with an exact amount, i.e. If Alice wants to send Bob 5MOB, but only has a single TxOut of value 10 MOB, she will have to first send 5 MOB to her own PublicAddress.